### DISCUSSION TITLE
XPlatformReleasenotes : Global WorkItems and commits returned

### DISCUSSION THREAD

**Discussion by egoux** [rfennell/AzurePipelines#1194] [Q&A]
Hello,

I'm using your excellent extension “XPlatformReleaseNotes” with Azure Devops Server 2019. 
My build pipeline is created in Classic mode with a TFVC source.

I'm having trouble with the results returned by this one.

When I map my project's main folder as a workspace the "**commits**" and "**workitems**" returned are less than real.

After investigation without success, I proceeded to a test by running a new build pipeline on each sub-folder of my project.
It seems that a limit of returning "**commits**" and "**workitems**" has been reached.
Is it upstream of the processing of the extension or related to the calls of:

```
globalCommits = yield buildApi.getBuildChanges(teamProject, buildId, "", 5000);
globalWorkItems = yield buildApi.getBuildWorkItemsRefs(teamProject, buildId, 5000);
```

PS :  Main Parameters are 

```
"getParentsAndChildren": "true",
"getAllParents": "true",
"SortWi": "true",
"SortCS": "false",
"overrideStageName": ""
"checkStage": "false"
```

You will find below the summary table of the "commits" and "work items" returned.

![image](https://user-images.githubusercontent.com/94963486/150317390-c90b4aff-95de-4bc2-9f7c-2e34667c95f0.png)

Best regards.
Eric

**Comment by rfennell** [comment]
We need to look at the logs in more details to see what is going on. There are two ways to do this

- Run a build with the diagnostic checkbox ticked at queue time, or set a variable `system.debug=true'
- Or use my [command line tester](https://github.com/rfennell/AzurePipelines/blob/main/Extensions/XplatGenerateReleaseNotes/V3/testconsole/readme.md) that allows you to re-run just my task - **I think this is a better option for diags**

The log will then show you all the calls being made and we can try to work out what is going on

**Reply by egoux** [reply]
Thanks
My example states a representation of the problem. The numerical results will be different on the logs.

I restarted the 5 build pipelines in debug=true mode.
Here the results for each : 

![image](https://user-images.githubusercontent.com/94963486/150331705-d98a224d-97c1-4261-a7e7-3d8206b8feb3.png)

Here are the attached files. 
[142_ALL_GenerateReleaseNotes.txt](https://github.com/rfennell/AzurePipelines/files/7904783/142_ALL_GenerateReleaseNotes.txt)

[143_GATEWAYS_GenerateReleaseNotes.txt](https://github.com/rfennell/AzurePipelines/files/7904784/143_GATEWAYS_GenerateReleaseNotes.txt)

[144_SERVICES_GenerateReleaseNotes.txt](https://github.com/rfennell/AzurePipelines/files/7904785/144_SERVICES_GenerateReleaseNotes.txt)

[145_SPHERES_GenerateReleaseNotes.txt](https://github.com/rfennell/AzurePipelines/files/7904786/145_SPHERES_GenerateReleaseNotes.txt)

[145_XPLATFORM_GenerateReleaseNotes.txt](https://github.com/rfennell/AzurePipelines/files/7904788/145_XPLATFORM_GenerateReleaseNotes.txt)

**Reply by rfennell** [reply]
After a quick look at the logs, I think the issue is that you are counting duplicated WI. 

For example WI 280 seems to appear as a directly associated WI in Builds 143, 144, 145, 146. It obviously also appears in 142, but only once. So this accounts for 3 of the difference in WI.

Remember the Azure DevOps API returns unique Changeset and WI associations. 

I don't think this is max return size issue is as the SDK calls I make will allow up to 5000 items on each call and we are well short of that

```
globalCommits = yield buildApi.getBuildChanges(teamProject, buildId, "", 5000);
globalWorkItems = yield buildApi.getBuildWorkItemsRefs(teamProject, buildId, 5000);
```

These SDK calls are just wrappers for the following REST calls, if you want to check the values. I am doing nothing to process or trim these lists
- [Changeset](https://docs.microsoft.com/en-us/rest/api/azure/devops/build/builds/get-changes-between-builds?view=azure-devops-rest-6.0)
- [Work Items](https://docs.microsoft.com/en-us/rest/api/azure/devops/build/builds/get-work-items-between-builds?view=azure-devops-rest-6.0)

**Comment by egoux** [comment]
I must have made a mistake in the sum, and my explanation is not the one I wanted to give you. I rephrase it differently

Take the example of Build 142 linked to ALL sources (GATEWAYS included) and build 143 linked to GATEWAYS source only.

I restarted the 2 generation tasks in local mode with your tester. I took a screenshot of the 2 output files produced.

![image](https://user-images.githubusercontent.com/94963486/150375377-3a3711b7-fbee-46bb-85dc-c89da8b84f27.png)

![image](https://user-images.githubusercontent.com/94963486/150375459-cc2070da-d99d-411e-98e9-c997f869dcda.png)



- Build 143 gives me the following WIs: 61, 98, 107,108, 277 and 280
- Build 142 does not give me the following WIs: 61, 98, 107 and 108

I'm troubled.

**Reply by rfennell** [reply]
What do you see if you use the REST API calls mentioned above directly? As they are just GETs you can call them in a browser, or use Postman or similar.

As I said I am doing no processing of the association, the global list arrays are just when the API returns

**Comment by egoux** [comment]
I saw you doing no processing of the association. 
That's what I feareds while studying your source codes.
I wrote a program in C# console and calling the REST API methodss : The problem is the same.

So, it's upstream of your extension task.

**Reply by rfennell** [reply]
As I feared, sorry there is nothing I can do about that

**Reply by egoux** [reply]
In any case, thank you for your collaboration and availability.

**Comment by egoux** [comment]
Hello,

I allow myself to come back to you because I found a ticket on "Microsoft Developer community" concerning the number of commits returned via the API. 
It seems that there is a problem and that its circumvention would go through the use of the "continuationtoken" parameter. 

`globalCommits = yield buildApi.getBuildChanges(teamProject, buildId, "", 5000);`

Here is the link to the ticket

https://developercommunity.visualstudio.com/search?space=22&q=commit+200&viewtype=solutions&entry=problem

![image](https://user-images.githubusercontent.com/94963486/150943545-4f86463c-3ceb-432e-8d2a-44981088dd3c.png)

**Reply by rfennell** [reply]
Node SDK  has the limit of 5000 items return (the 5000 parameter), as does the REST API without the need to manage the continuation token. You are well under this limit so I don't think it is your problem

**Reply by egoux** [reply]
I opened a ticket on Microsoft developper community. Because I have read other testimonies that the top parameter was not always interpreted. 

I will keep you informed.

**Reply by egoux** [reply]
I did a test with continuationtoken. 
The problem is the same. 
Stuck at 201, and if I create a new commit it goes into the list and the oldest in the list disappears.

**Reply by rfennell** [reply]
TOP=5000 works for me

**Reply by egoux** [reply]
TOP works for me too but until 201.
The method : `getBuildChanges(teamProject, buildId, "", 5000);` only reads a **build constituted upstream**. 
So the problem is on the TFVC build pipeline not in your extension.

**Reply by rfennell** [reply]
I would not 'hold your breath' for TFVC changes/fixes, the focus does seem to be very much on Git these days

**Reply by egoux** [reply]
I know, that's what worries me. 
And also for historical reasons and the density of our solution. We don't have the human and time resources.
So wait and see.
